### PR TITLE
All the data stuffs

### PR DESCRIPTION
**Pull request by eliascodes**
Related to #14 
Related to #11 
Related to #6

Apologies for the crap code, and the long wait.

Also, this is a semi WIP, in that some files still need to be commented and tested, but there will be no further changes to code.

The code is now broadly separated into extract/transform/load stages, as should be clear from the directory structure.

Let's talk it through together at some point, either before or after you get a chance to leave comments, as you prefer.

### CODE REVIEW

File: src/extract/visitor/fieldMap.js [hunk 1]
```diff
@@ -28,38 +28,42 @@ module.exports = {
   },
   users: {
     id: 'user_id',
+    cb_id: 'fk_user_to_organisation',
     fullname: 'user_name',
     sex: 'user_gender',
     yearofbirth: 'user_birth_year',
-    hash: 'user_secret',
+    hash: 'user_qr_code',
```
Comment (astroash) [inline src/extract/visitor/fieldMap.js]: named `qr_code` on the remote db schema

File: src/extract/visitor/fieldMap.js [hunk 2]
```diff
@@ -28,38 +28,42 @@ module.exports = {
...
-    tuesday: 'activity_tuesday',
-    wednesday: 'activity_wednesday',
-    thursday: 'activity_thursday',
-    friday: 'activity_friday',
-    saturday: 'activity_saturday',
-    sunday: 'activity_sunday',
-    date: 'activity_created_at',
+    id: 'visit_activity_id',
+    cb_id: 'fk_visit_activity_to_organisation',
+    name: 'visit_activity_name',
```
Comment (astroash) [inline src/extract/visitor/fieldMap.js]: Should these column names match the online db schema?

File: src/load/organisation.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ */
+const {
+  map, head, filter, compose, pickAll,
+} = require('ramda');
+const { sanitiseEntity, mapKeys } = require('../shared/util');
+
+const isFloat = (n) => n && !Number.isNaN(Number.parseFloat(n));
+
+const sanitiseOrg = compose(
+  ({ lat, lng, ...o }) =>
```
Comment (astroash) [inline src/load/organisation.js]: this would be more readable if this anonymous function is pulled out

File: src/transform/pre_process.js [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+  entities.user.forEach((user) => {
+    // User role mapping
+    // > Users coming from the admin app have role names corresponding directly to
+    //   roles in the new data model
+    // > Users coming from the visitor app have no role names but are all by definition
+    //   visitors
+    if (typeof user.user_role_name === 'string') {
+      switch (user.user_role_name) {
+        case 'System Admin':
+          user.user_role_name = 'SYS_ADMIN';
```
Comment (astroash) [inline src/transform/pre_process.js]: Umm... is this using a forEach to mutate the nested values of an object?

File: .travis.yml [hunk 5]
```diff
@@ -13,14 +13,15 @@ install:
   - npm install -g codecov
   - npm install
 
+before_script:
+  - echo "module.exports = {};" > ./src/transform/region_map.js
 
 script:
   - npm run lint
   - npm run cover
-  - npm run cover:report
```
Comment (astroash) [inline .travis.yml]: This script can also be removed from the `package.json`

File: src/extract/visitor/fieldMap.js (hunk 2 above)
Comment (eliascodes) [reply src/extract/visitor/fieldMap.js]: No

File: src/extract/visitor/fieldMap.js (hunk 1 above)
Comment (eliascodes) [reply src/extract/visitor/fieldMap.js]: This name doesn't correspond to the target schema.

File: bin/cli.js [hunk 6]
```diff
@@ -1,11 +1,20 @@
 #!/usr/bin/env node
+require('dotenv').config();
+const { mergeDeepRight } = require('ramda');
 const parser = require('minimist');
 const migrate = require('../src');
 
 const main = (argv) => {
-  const { visitorSource, adminSource, target } = parser(argv.slice(2));
+  const defaults = {
```
[the remark this answers is not in the document]
Comment (eliascodes) [reply bin/cli.js]: They're environment variables, so they are by definition strings.

File: src/transform/pre_process.js (hunk 4 above)
Comment (eliascodes) [reply src/transform/pre_process.js]: Yes 😬

**Comment by eliascodes** [comment]
I'm not going to add more tests to get it back above 97% coverage or whatever it was, so if the tests pass and you don't have any more comments, you can go ahead and merge it.

**Pull request merged**